torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -1199,6 +1200,7 @@ class AsyncSparseAllreduceCUDAWork : public AsyncSparseAllreduceWork {
     for (size_t i = 0; i < inputs.size(); i++) {
       stream_guard.reset_stream(streams[i]);
       outputs.push_back(output.to(inputs[i].device(), /*non_blocking=*/true));
+      inputs[i].copy_(output, /* non_blocking */ true);
```

Comment:
how does non_blocking work for sparse tensors? For dense CPU tensors, we will need to use pinned memory to make it non_blocking, but the in-place copy for sparse tensors could change the number of elements, which might change the storage, which might prevent us from pinning the memory??

Reply:
hmm, I'm not sure about pinning the memory underneath. 

But output.to(inputs[i].device(), /*non_blocking=*/true) is also kind of sparse tensor non_blocking copy?

For safety, I will change this line to "inputs[i].copy_(outputs[i], /* non_blocking */ true);"

Reply:
It's just me don't understand how non_blocking works for sparse tensors. If it is dense and CPU to CPU, IIUC, non_blocking is ignored. And `non_blocking` only works between CPU and CUDA.

Yep, for now, removing non_blocking should be OK.

Reply:
I moved this copy into synchronize() call, please see updated PR. because we have to copy btw devices only after previous non blocking copy call is finished. 

regarding how non_blocking works for sparse tensors, I searched a little bit and did not find how it works with pinned memory, but from the #9005, looks like indices, values are copied separately and they have to be coalesced first